spec/p2p/reactor/reactor.qnt

```diff
@@ -0,0 +1,238 @@
+
+		// The reactor runs multiple routines.
+		routines: Set[Routine],
+
+		// Service implementation.
+		started: bool,
+		stopped: bool,
+
+		// Indicates whether a switch was set: reactor registration.
+		switch: bool,
+	}
+
+	var state: ReactorState
+
+	/*
+	 * GetChannels returns the list of MConnection.ChannelDescriptor. Make sure
+	 * that each ID is unique across all the reactors added to the switch.
+	 */
+	pure def GetChannels(s: ReactorState): Set[ChannelDescriptor] = {
+		// This method should return a list of channels, e.g.:
+		Set({ID: 3, Priority: 1})
+		.union(Set({ID: 7, Priority: 2}))
+	}
+
+	/*
+	 * SetSwitch allows setting a switch.
+	 */
+	pure def SetSwitch(s: ReactorState, switch: bool): ReactorState = {
+		s.with("switch", switch)
+	}
+
+	// Code: Switch.AddReactor(name string, reactor Reactor)
+	action register() = all {
+		not(state.switch),
+		val channels = state.GetChannels()
+		val chIDs = channels.map(c => c.ID)
+		channels.size() == chIDs.size(), // channel IDs are unique
+		state' = state.SetSwitch(true),
+	}
+
+	/*
+	 * Start the service.
+	 * If it's already started or stopped, will return an error.
+	 */
+	pure def OnStart(s: ReactorState): ReactorState = {
+		// Startup procedures should come here.
+		s.with("started", true)
+	}
+
+	// Code: Switch.OnStart()
+	action start(): bool = all {
+		state.switch,
+		not(state.started) and not(state.stopped),
+		state' = state.OnStart()
+	}
+
+	/*
+	 * Stop the service.
+	 * If it's already stopped, will return an error.
+	 */
+	pure def OnStop(s: ReactorState): ReactorState = {
+		// Shutdown procedures should come here.
+		s.with("stopped", true)
+	}
+
+	// Code: Switch.OnStop()
+	action stop(): bool = all {
+		state.started and not(state.stopped),
+		// Either no peer was added or all peers were removed
+		state.peers.size() == 0,
+		state' = state.OnStop()
+	}
+
+	/*
+	 * Return true if the service is running.
+	 */
+	pure def IsRunning(s: ReactorState): bool = {
+		s.started and not(s.stopped)
+	}
+
+	/*
+	 * InitPeer is called by the switch before the peer is started. Use it to
+	 * initialize data for the peer (e.g. peer state).
+	 */
+	pure def InitPeer(s: ReactorState, peer: Peer): (ReactorState, Peer) = {
+		// This method can update the received peer, which is returned.
+		val updatedPeer = peer
+		(s.with("peers", s.peers.union(Set(peer.ID))), updatedPeer)
+	}
+
+	// Code: Switch.addPeer(p Peer): preamble
+	action initPeer(peer: Peer): bool = all {
+		state.switch and state.IsRunning(),
+		not(peer.ID.in(state.peers)),
+		state' = state.InitPeer(peer)._1 // Updated peer not used here
+	}
+
+	/* 
+	 * AddPeer is called by the switch after the peer is added and successfully
+	 * started. Use it to start goroutines communicating with the peer.
+	 */
+	pure def AddPeer(s: ReactorState, peer: Peer): ReactorState = {
+		// This method can be used to start routines to handle the peer.
+		// Below an example of an arbitrary 'ioRoutine' routine.
```

Comment:
Where do the available routines come from? Maybe add them in a set in the ReactorState and them them here and then start.
Are they all started together? If so, start all here. Otherwise, chose one in addPeer and pass as a parameter to AddPeer

Reply:
This is an abstraction, we don't intend here to represent the operation of a specific reactor. In fact, each reactor is free to do whatever it wants at this method and each reactor has its own set of routines. The idea here is to represent that this is _right_  point where to start routines. Routines should not be started at `InitPeer`, as the `Peer` passed to that method is not started yet.

Reply:
> Otherwise, chose one in addPeer and pass as a parameter to AddPeer

We want to preserve the same parameters passed to the interface `Reactor` methods. The method only receives a `Peer` on the API, so we want to reproduce the same here.

Reply:
More broadly, in a next refinement/extension of the specification we might represent routines interacting with peers. These routines, for instance, send messages to a peer using the `Peer` handler. For now routines are very simpler, but we might make them more "useful".

Reply:
Got it. I just found it weird to have these examples as part of the spec. Then again, I don't know how to provide examples in a better way.
I wish we could define just the prototype here, for type checking, and force a refined instantiation when testing.